File: sin_gen.py
from numpy import *;
from matplotlib.pyplot import *;
import scipy.linalg;
from math import sin;

# Input, reservoir, output size
res_size = 20;
in_size = 0;
out_size = 1;

# Training data size and washout time
test_size = 50;
train_size = 300;
washout_size = 100;

# Scaling parameters
w_scale = 0.98;
w_back_scale = 1.01;
bias_scale = 0;
reg = 1e-7;

# Compute w
random.seed(0);
w_0 = random.uniform(-1.0, 1.0, (res_size, res_size));
spectral_rad_w_0 = max(abs(linalg.eig(w_0)[0]));
w = w_0 * 1.0/spectral_rad_w_0 * w_scale;

# Compute w_back
w_back = random.uniform(-1.0, 1.0, (res_size, 1)) * w_back_scale;

# Compute bias
bias = random.normal(size = (res_size, 1)) * bias_scale;

# Teaching signal
d = [0.5 * sin(i / 4) for i in range(0, train_size + test_size + 1)];

# Drive network using teaching signal
m = zeros((train_size - washout_size, res_size));
t = resize(d[washout_size:train_size], (train_size - washout_size, 1));
# Reservoir
x = zeros((res_size, 1));
internal_training = [];
for i in range(train_size):
	x_old = x;
	x = tanh(dot(w, x) + dot(w_back, d[i]) + bias);
	internal_training.append(x[1:5,0]);
	if(i >= washout_size):
		m[i - washout_size] = x_old[:,0];

# w_out = dot(linalg.pinv(m), t);
w_out = dot(linalg.inv(dot(m.T,m) + reg * eye(res_size)), dot(m.T, t));
print("Abs Mean of w_out",mean(abs(w_out)));

y = zeros(test_size);
u = d[train_size]
internal_generating = [];
for i in range(test_size):
	x = tanh(dot(w, x) + dot(w_back, u) + bias);
	internal_generating.append(x[1:5,0]);
	y[i] = dot(w_out.T, x);
	u = y[i];
print("MSE =", 
	sum(square(d[train_size + 1 : train_size + test_size + 1] - y[0:test_size]))
	/ test_size);

figure(1);
subplot(3, 1, 1);
plot(internal_training);
subplot(3, 1, 2);
plot(internal_generating);
subplot(3, 1, 3);
plot(y);
plot(d[train_size + 1 : train_size + test_size + 1]);
show();

File: sin_gen.m
clear all;
clc;

% Input, reservoir, output size
res_size = 20;
in_size = 0;
out_size = 1;

% Training data size and washout time
test_size = 50;
train_size = 300;
washout_size = 100;

% Scaling parameters
w_scale = 0.98;
w_back_scale = 1.01;
bias_scale = 0;
reg = 1e-8;

rand('seed', 42);
% Scale internal weights w
w_0 = 2 * rand(res_size, res_size) - 1;
spectral_rad_w_0 = max(abs(eigs(w_0)));
w = w_0 * ((1.00 / spectral_rad_w_0) * w_scale);

% Compute w_back
w_back = (2 * rand(res_size, out_size) - 1) * w_back_scale;

% Compute bias
bias = (2 * rand(res_size, out_size) - 1) * bias_scale;

% Compute teacher forcing vector
sample_points = 1:(train_size + test_size + 1);
d = sin(2 * pi * sample_points/20);

% Internal state collector m and teacher collector t
m = zeros(train_size - washout_size, res_size);
t_T = transpose(d);
t = t_T(washout_size + 1:train_size,:);

x = zeros(res_size, 1);
u = 0;
for i = 1:train_size
    x = tanh(w * x + w_back * u + bias);
    u = d(i);
    if i > washout_size
        m(i - washout_size, :) = transpose(x);
    end
end

% Compute output weights
w_out = (inv((transpose(m) * m) + reg .* eye(res_size)) * transpose(m) * t);

% Compute mean_abs and mse_train
mean_abs = mean(abs(w_out));
mse_train = mse(w_out' * m', t');

y = zeros(1,test_size);
internal_testing = zeros(test_size, res_size);
for i = 1:test_size
    x = tanh(w * x + w_back * u + bias);
    internal_testing(i,:) = transpose(x);
    y(i) = transpose(w_out) * x;
    u = y(i);
end

% Compute mse_test
mse_test = mse(y, d(train_size + 1:train_size + test_size));

figure;
% subplot(3,1,1);
% plot(m(:,1:5));
% subplot(3,1,2);
% plot(internal_testing(:,1:5));
% subplot(3,1,3);
p = plot(1:test_size, y, 'c', 1:test_size, d(train_size + 1:train_size + test_size), 'r--');
p(1).LineWidth = 4;